### PR TITLE
Implement differentiable CGP

### LINKED ISSUE #59: Provide mechanisms to combine evolution with local search
In particular a mechanism for the updating of node values from pytorch classes in which parameters have been optimized via gradient descent would be desirable. #57 provides a first stab at this.

### PR DESCRIPTION
**Pull request by jakobj**
- add a new node CGPParameter that, in contrast to the ConstantFloat can store arbitrary values
- these parameter values are stored in each individual and passed on to their offspring
- parameters can be updated from torch classes generated from the graph; this way parameters can be optimized with local search and then written back to the individuals
- add a new function template to provide local search via torch optimizers

as discussed. rebased onto master and added some improvements. this supersedes #57.

closes #59

### CODE REVIEW

**Comment by jakobj** [comment]
i introduced the gradient clipping (https://github.com/jakobj/python-gp/blob/57299e9a9b34ab9dd2ac6642ddea064f5b672cb5/gp/local_search/gradient_based.py#L50) since i've experienced a few issues with exploding gradients. however i think this makes sense in general: as the parameters are passed on to offspring we do not want them to be tuned to ridiculous values if the graph is completely wrong. instead we want them to be adapted in parallel to the graph over many generations.
i am not sure whether we want to include this explicitly like this or let users include it if they need it. this would require making the `gradient_based` function a bit more flexible. what do you think @mschmidt87 ?

**Comment by jakobj** [comment]
there's another interesting problem: the current implementation relies on `Parameter` nodes that provide numeric values to the graph. due to the logic of CGP these nodes can be used more than once in the final expression. for example it can happen that a node is used with opposite signs  in an expression as in `f(x) = -c * x** 2 + c` if the final expression is something like `f(x) = -3 * x ** 2 + 1` and the current value of c is 2.0 the gradient will pull in opposing directions, essentially halting the search although the structure of the expression is already correct. other approaches, e.g., Izzo et al 2017, consider multiplicative constants in each node instead. these are private to each node an hence can not lead to this effect.
one could consider an implementation where each node has a multiplicative and an additive constant that by default are one and zero, that could be tuned via local search. this would make the `Parameter` node proposed here obsolete. what do you think @mschmidt87 ?

File: examples/example_differential_evo_regression.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    ax_function.set_xlabel(r"$x$")
+
+    history, champion = evolution()
+
+    print(f"Final expression {champion.to_sympy()[0]} with fitness {champion.fitness}")
+
+    history_fitness = np.array(history["fitness_parents"])
+    ax_fitness.plot(np.max(history_fitness, axis=1), label="Champion")
+    ax_fitness.plot(np.mean(history_fitness, axis=1), label="Population mean")
+
```
Comment (mschmidt87) [inline examples/example_differential_evo_regression.py:131]:
```suggestion
ax_fitness.set_yscale('symlog')
```

File: gp/individual.py [hunk 2]
```diff
@@ -129,7 +149,33 @@ def to_sympy(self, simplify=True):
...
+        torch_cls : torch.nn.module
+            Instance of a torch class.
+
+        Returns
+        -------
+        None
+
+        """
+        for name, value in torch_cls.named_parameters():
+            name = f"<{name[1:]}>"
```
[the remark this answers is not in the document]
Comment (jakobj) [reply gp/individual.py:171]: nothing, they are meant literally as in the string should be something like the string "<p1>". we also cut of the leading underscore "_p1" -> "<p1">

File: gp/local_search/gradient_based.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    clip_value : float, optional
+        Clipping value for gradients. Clipping is skipped when set to
+        None. Defaults to 1e4.
+
+    """
+    if torch is None:
+        raise ModuleNotFoundError("No module named 'torch' (extra requirement)")
+
+    if optimizer is None:
+        optimizer_class = torch.optim.SGD
```
[the remark this answers is not in the document]
Comment (jakobj) [reply gp/local_search/gradient_based.py:35]: i knew i should've commented on this ;) we can't do that bc we first need to check in line 31 whether torch is available. otherwise it would make sense as you suggest.

**Comment by jakobj** [comment]
thanks for the review! i've addressed your comments. i've also made some minor changes in the parameters and introduced a new default for `clip_values`. please have another look.

File: gp/individual.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mschmidt87) [reply gp/individual.py:171]: Got it.

**Comment by mschmidt87** [comment]
> one could consider an implementation where each node has a multiplicative and an additive constant that by default are one and zero, that could be tuned via local search. this would make the `Parameter` node proposed here obsolete. what do you think @mschmidt87 ?

I like that suggestion. However, what about first merging this and then making this a feature of the next release? We could add a warning in the documentation for the user to raise awareness of the problem with the gradients? Or do you think this is a deal-breaker?

File: gp/local_search/gradient_based.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jakobj) [reply gp/local_search/gradient_based.py:38]: sure, we could. i just handled it like this bc we do it similarly in other parts, e.g., the tests. i guess the question is whether this remains the only function in this module or not. also, i'm not sure when to raise the error if we put the check in the `__init__`? when the users tries to access `gradient_based`?

**Comment by jakobj** [comment]
agreed, no need to solve everything in one PR. the current implementation is good for now, but yes, it has some algorithmic problems. let's address those in a later version. i don't think we need documentation on this. this sounds more like research than documentation ;)

File: gp/local_search/gradient_based.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mschmidt87) [reply gp/local_search/gradient_based.py:38]: I mean, if `torch` is not available, this module should not be available. If you do something like this in `__init__.py` this should work, right?
```python
try:
    import torch  # noqa: F401
    from .gradient_based import gradient_based
except ModuleNotFoundError:
    pass

[the remark this answers is not in the document]
Comment (jakobj) [reply gp/local_search/gradient_based.py:38]: yes, of course. but how do we handle the case where torch is not available and the user tries to `local_search.gradient_based`. they would just get a noninformative "ModuleNotFoundError", right? the way it's currently implemented they will have the information that this exists, but they need the extra requirement to use it.

[the remark this answers is not in the document]
Comment (mschmidt87) [reply gp/local_search/gradient_based.py:38]: Okay, I see your point. What about doing this then:
```python

try:
    import torch
except ModuleNotFoundError:
    torch = None

if torch is None:
    def gradient_based(*args, **kwargs):
          raise ModuleNotFoundError("No module named 'torch' (extra requirement)")
else:
    def gradient_based(..., optimizer=torch.optim.SGD, ...):
        ## full function definition
```

**Comment by mschmidt87** [comment]
Sounds good to me.

**Comment by mschmidt87** [comment]
As far as I can see, just the one point about availability of torch is remaining.

But one more thing: Can you add some description to the README about this?

File: gp/local_search/gradient_based.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jakobj) [reply gp/local_search/gradient_based.py:38]: interesting, haven't thought of that :thinking: do you think this is easier to understand than just accepting a `None` as a default? or are you afraid this is going to lead to problems with mypy? besides the seeemingly increased complexity, the additional indentation bugs me

[the remark this answers is not in the document]
Comment (mschmidt87) [reply gp/local_search/gradient_based.py:38]: I think it is easier to understand, but I get the point with the additional indent. Maybe it's fine to leave it as it is.

**Comment by mschmidt87** [comment]
I'm fine with the code changes, just add some documentation to the README please.

**Comment by jakobj** [comment]
ok, i've updated the README, please have another look @mschmidt87

File: README.md [hunk 4]
```diff
@@ -13,96 +15,75 @@ This library implements Cartesian Genetic Programming (e.g, Miller and Thomson,
 
 A simple example of CGP applied to a symbolic regression problem can be found in `examples/example_evo_regression.py`.
 
+Searching numerical values for constants via local search
+---------------------------------------------------------
+
+This library supports the use of constants as operator nodes in order to represent expressions like `f(x) = x + 1.0`.
```
Comment (mschmidt87) [inline README.md:21]: better use `c` in the equation instead of `1.0`.

**Comment by jakobj** [comment]
good point, fixed it now. if you're happy pls approve so that i can merge ;)

**Comment by jakobj** [comment]
travis and you are happy. using my administrator power to merge :smiling_imp:
